Fc 1146 event sources (#7)
Templum/developers-community: pages/Documents/DeveloperTools/LivePersonFunctions/EventSources/automatic-messages.md

---
pagename: Automatic Messages
keywords:
sitesection: Documents
categoryname: Developer Tools
documentname: LivePerson Functions
subfoldername: Event Sources
permalink: liveperson-functions-event-sources-automatic-messages.html
indicator: both
---

Automatic Messages (also known as System Messages) are predefined messages triggered upon events that occur through the lifecycle of a conversation (e.g. the consumer opens a new conversation, the conversation is transferred to another agent, the time to respond is updated, etc.). Auto messages will be sent to both agent and consumer by the system. Therefore they appear in the conversation’s history both on the consumer and conversation cloud side.

Thanks to our integration with Automatic Messages we are able to have functions being triggered based on selected events. During the invocation, a payload is sent to a function containing metadata related to the specific conversation and based on the selected event. This payload can then be used in the function for further processing and referencing.

<img src="img/functions/functions_automessages_flow.png" alt="Functions: Automatic Messages Flow" style="width:100%;"/>

{: .notice}

> If an specific system message is disabled, the associated function will be invoked regardless of the state of the enabled flag. In order to completely avoid firing of functions you can either undeploy your function or Ignore it in your function code. You can distinguish the system message with the `cbotEventType` property included in the invocation payload.

Reviewer: We need to be careful, ignoring won't stop the firing. So it would be still counted as an invocation, ignoring a event is optional but would still trigger an invocation

```suggestion
If a specific system message is disabled, the associated function will be invoked regardless of the state of the enabled flag. In order to completely avoid firing of functions, you have to undeploy your function. You may choose to ignore a specific event by leveraging the `cbotEventType` property, that is included in the invocation payload. However, the ignored invocations here would still count towards your usage limits.
```

<tr><td>SYSTEM_END_CONVERSATION</td><td>Messaging conversation end</td></tr>
<tr><td>ENGAGEMENT_TRANSFERED</td><td>Messaging conversation routing</td></tr>
<tr><td>YOU_ARE_CONNECTED_TO</td><td>Messaging participants change</td></tr>
<tr><td>BACK_TO_QUEUE</td><td>Messaging participants change</td></tr>
<tr><td>JOINED_AGENT_LEFT_CONVERSATION</td><td>Messaging participants change</td></tr>
<tr><td>ANOTHER_AGENT_JOINED_CONVERSATION</td><td>Messaging participants change</td></tr>
<tr><td>AGENT_UPDATED_RESPONSE_TIME</td><td>Messaging TTR</td></tr>
<tr><td>CONSUMER_CANCELED_URGENT_RESPONSE_TIME</td><td>Messaging TTR</td></tr>
<tr><td>CONSUMER_ASKED_URGENT_RESPONSE_TIME</td><td>Messaging TTR</td></tr>
<tr><td>CONSUMER_SEND_MESSAGE_ON_OFFHOURS</td><td>Messaging Line in Off-Hours</td></tr>
<tr><td>CONV_AWAITING_IN_QUEUE</td><td>Messaging conversation idle</td></tr>
<tr><td>AGENT_NON_RESPONSIVE</td><td>Messaging conversation idle</td></tr>
<tr><td>CONSUMER_NON_RESPONSIVE</td><td>Messaging conversation idle</td></tr>
<tr><td>CONSUMER_STEP_UP</td><td>Messaging consumer step up</td></tr>
</tbody></table>

### Interaction via Commands

> You have the option to send commands back to the trigger. You have the option to send one command or multiple commands at a time, by simply adding them to an array. Using callback commands is **not** mandatory. If you only wish to use the events listed above to trigger functions and nothing else, there's no reason for you to pass callback commands back, we recommend you to pass and empty array (`[]`) instead.

Reviewer: ```suggestion
You have the option to send commands back to the trigger. Where you can choose to send one command or multiple commands at a time, by simply adding them to an array. Using callback commands is **not** mandatory. If you don't want to take any actions, simply return an empty array `[]` this will result in no operation.
```

```javascript
let result = [
   {
       type: "systemMessage", // Returns a system message into the conversation
       text: "your message"
   },
   {
       type: "transfer", // Transfers the conversation.
       skillId: "123456", // Transfer to different skill.
       agentId: "123456" // Propose an agent.
   },
   {
       type: "closeConversation" // Closes the conversation
   }
]
callback(null, result);
```

With the Automatic Messages as the trigger, you have the option to execute the following commands:

* Send a message. The default automatic message can be overwritten with:

```json/js
{
            type: "systemMessage",
            text: "your message"
}
```

> To set the message empty just pass `[]` as text.

Reviewer: Can't they simply set empty string aka '' ? Honesty that makes much more sense than [], which just stringifies into an empty string ....

```suggestion
To set the message empty just pass `[]` as text.
```

Reviewer: And if thats the case, we don't need to tell them. People should have that level of understanding anyways

Author: i have reomved

{: .important}

> If no message is set in the result of the function (which it returns to the trigger, for example: `callback();` ), the default automatic message for the account will be triggered.

Reviewer: This should go before all the coding samples.

Reviewer: Especially since it touches on our suggestion to provide [] for NO-OP

Author: done

* Transfer Conversation to a different Skill:

```javascript
let result = [
   {
       type: "transfer", // Transfers the conversation.
       skillId: "123456", // Transfer to different skill.
   }
]
callback(null, result);
```

* Transfer Conversation to a different Agent (to configure this feature see [here](https://knowledge.liveperson.com/contact-center-management-messaging-operations-transfer-to-agent.html)):

```javascript
let result = [
   {
       type: "transfer", // Transfers the conversation.
       agentId: "123456", // Transfer to different skill.
   }
]
callback(null, result);
```

* Close the Conversation:

```javascript
let result = [
   {
       type: "closeConversation" // Closes the conversation
   }
]
callback(null, result);
```

{: .notice}
If you add more than one command of a certain type (e.g. 2 messages) **only the first command** of this type will be processed.

### Best Practices

#### Messaging event are asynchronous

> Functions for messaging listens for messaging events asynchronously. As a consequence, this can cause race conditions with other parts of the conversation cloud. Therefore, it is considered best practice to use bots instead of Functions for implementing routing logic. Functions are a good option to enrich data with third-party systems like CRMs or to save data in the [Conversation Context Service](maven-context-warehouse-overview.html) in order to make it usable by another system like the Conversation Orchestrator. Routing via Functions makes sense whenever a conversation is in a stagnant state (i.e. not in process of being routed), e.g. a conversation is idle or a message line has been sent in off-hours. Otherwise, the asynchronous nature of its events might interfere with the proper flow of a "dynamic" conversation.

Reviewer: ```suggestion
Functions for messaging listens for messaging events asynchronously. As a consequence, this can cause race conditions with other parts of the conversation cloud. Therefore, it is considered best practice to use bots instead of Functions for implementing routing logic. Routing via Functions makes sense whenever a conversation is in a stagnant state (i.e. not in process of being routed), e.g. a conversation is idle or a message line has been sent in off-hours. Functions are a good option to enrich data with third-party systems like CRMs or to save data in the [Conversation Context Service](maven-context-warehouse-overview.html) in order to make it usable by another system like the Conversation Orchestrator.
```

{: .notice}
if the authenticated consumer's previous conversation was auto-closed, and the new one opened within 48 hours of that, the new conversation event won't be triggered.

#### Automatic Messages Error Handling

If your function invocation takes too long or you have coding errors, Automatic Messages will retry up to 3 times, this will increase the number of invocations when your function does not perform well and could lead to more overload.

### Payload Details

<table>
<thead><tr><th>1. level</th><th>2. level</th><th>3. level</th><th>description</th><th>type</th><th>example</th></tr></thead><tbody>
 <tr><td>end</td><td>closeReason</td><td>&nbsp;</td><td>which role closed conversation</td><td>STRING</td><td>AGENT/CONSUMER/SYSTEM</td></tr>

>  <tr><td>general</td><td>cbotEventType</td><td>&nbsp;</td><td>ID of the system message</td><td>STRING</td><td>CONSUMER_OPEN_NEW_CONVERSATION_FIRST_TIME_OFF_HOURS</td></tr>

Reviewer: ```suggestion
 <tr><td>general</td><td>cbotEventType</td><td>&nbsp;</td><td>Original conversation event type</td><td>STRING</td><td>CONSUMER_OPEN_NEW_CONVERSATION_FIRST_TIME_OFF_HOURS</td></tr>
```

---
pagename: Automatic Messages
keywords:
sitesection: Documents
categoryname: Developer Tools
documentname: LivePerson Functions
subfoldername: Event Sources
permalink: liveperson-functions-event-sources-automatic-messages.html
indicator: both
---

Automatic Messages (also known as System Messages) are predefined messages triggered upon specific events that are detected by the system (e.g. the consumer opens a new conversation, the conversation is transferred to another agent, the time to respond is updated, etc.). When the auto messages are sent, they are displayed to the consumer and the agent within the conversation transcript, and they also appear in the conversation’s history both on the consumer’s side and in LiveEngage.

> With Functions for messaging you are able to invoke `lambdas` from standard messaging events thanks to our integration with Automatic Messages. Along with the invocation, the function is sent a payload containing metadata related to the conversation which invoked the function. This payload can then be used in the function for further processing and referencing.

Reviewer: ```suggestion
Thanks to our integration with Automatic Messages we are able to have functions being triggered based on selected events. During the invocation, a payload is sent to a function containing metadata related to the specific conversation and based on the selected event. This payload can then be used in the function for further processing and referencing.
```

<img src="img/faas-automessages-flow.png" alt="FaaS Automatic Messages Flow" style="width:70%;"/>

> <div class="important">If an specific system message is disabled, the associated function will be invoked regardless of the state of the enabled flag.</div>

Reviewer: In situation where you want to warn the user, please prefer notice block over this div approach
```suggestion
<div class="important">If an specific system message is disabled, the associated function will be invoked regardless of the state of the enabled flag.</div>
```

Author: you mean with this: {: .important}?

Reviewer: {: .notice}

This is the Automatic Messages flow with Functions integration:

1. Controller Bot listens at conversation events.
2. If there is an Auto-message configured for the event --> It will send that message to the conversation.
3. If there is a Function deployed associated to that event -->  The function will be invoked.
4. After invocation --> Controller Bot will check the response and will perform one of the following actions based on the callback payload:
    1. Transfer Conversation to a different Skill.

>     2. Transfer Conversation to a different Agent.

Author: forgot

>     2. Transfer Conversation to a different Agent.

Author: done

    3. Close the Conversation.
    4. Do nothing.

### Configuration

#### Step 1 - Enable Automatic Messages

Enable the Automatic Messages feature.

<div class="important"> It is required that your account has the Controller Bot permissions enabled; please contact your account team in order to do this.</div>

#### Step 2 - Create a function

Create a new function using one of the following messaging events:

* Messaging new conversation.
* Messaging conversation end.
* Messaging conversation routing.
* Messaging participants change.
* Messaging TTR.
* Messaging Line in Off-Hours
* Messaging conversation idle.
* Messaging consumer step up.

> You can select one of the available [templates](liveperson-functions-event-sources-overview.html#templates) for the chosen event. Currently, only one function per template type can be created for those conversational events. If there are multiple types of functionality needed that stem from the same event invocation, these should be coded into the same `lambda`.

Reviewer: ```suggestion
You can select one of the available [templates](liveperson-functions-event-sources-overview.html#templates) for the chosen event. Currently, only one function per template type can be created for those conversational events. If there are multiple types of functionality needed that stem from the same event invocation, these should be coded into the same `lambda`.
```

Here we could make advertisement for our orchestrator functions

You can select a Skill or set of Skills. Specifying a skill has the advantage of reducing the complexity of the lambda's code. Furthermore, it reduces the number of unnecessary invocations.

#### Step 3 - Edit your Function

> Adjust the coding from the template according to your needs by modifying the function. The function can return a series of commands back to the invoker.

Reviewer: Linkt to the relevant Deep Dive CLI/UI section here

Author: Done check is you like

#### Step 4 - Deploy your function

Just like any other function, this function must be deployed before it can be used. [Please see this document](liveperson-functions-getting-started-your-first-function.html#deploy) for more information on how to deploy your function. At this point, you can also test your function.

> 

Reviewer: Linkt to the relevant Deep Dive CLI/UI section here

Reviewer: Not the first function guide !

Author: done

<tr><td>AGENT_UPDATED_RESPONSE_TIME</td><td>Messaging TTR</td></tr>
<tr><td>CONSUMER_CANCELED_URGENT_RESPONSE_TIME</td><td>Messaging TTR</td></tr>
<tr><td>CONSUMER_ASKED_URGENT_RESPONSE_TIME</td><td>Messaging TTR</td></tr>
<tr><td>CONSUMER_SEND_MESSAGE_ON_OFFHOURS</td><td>Messaging Line in Off-Hours</td></tr>
<tr><td>CONV_AWAITING_IN_QUEUE</td><td>Messaging conversation idle</td></tr>
<tr><td>AGENT_NON_RESPONSIVE</td><td>Messaging conversation idle</td></tr>
<tr><td>CONSUMER_NON_RESPONSIVE</td><td>Messaging conversation idle</td></tr>
<tr><td>CONSUMER_STEP_UP</td><td>Messaging consumer step up</td></tr>
</tbody></table>

### Callback commands

You have the option to send callback commands back to the invoker. You can add multiple commands to the response as the functions itself can return an array or a single object.

With the controller bot as the invoker, as is the case for messaging events, you have the option to execute the following callback commands:

* Send a message. The default automatic message can be overwritten with:

<code>{
            type: "systemMessage",
            text: "your message"

>         }</code>. To set the message empty just pass <code>[]</code> as text.

Reviewer: Please ```json/js instead of <code> and `` over <code> when inlined....

Overall **avoid** HTML inside markdown

       agentId: "123456", // Transfer to different skill.
   }
]
callback(null, result);
```

* Close the Conversation:

```javascript
let result = [
   {
       type: "closeConversation" // Closes the conversation
   }
]
callback(null, result);
```

Here's an example of a response sent back to the invoker using multiple of these commands:

```javascript
let result = [
   {
       type: "systemMessage", // Returns a system message into the conversation
       text: "your message"
   },
   {
       type: "transfer", // Transfers the conversation.
       skillId: "123456", // Transfer to different skill.
       agentId: "123456" // Propose an agent.
   },
   {
       type: "closeConversation" // Closes the conversation
   }
]
callback(null, result);
```

<div class="important">Using callback commands is <b>not</b> mandatory. If you only wish to use the events listed above to trigger functions and nothing else, there's no reason for you to pass callback commands back.</div>

If you add more than one command of a certain type (e.g. 2 messages) **only the first command** of this type will be processed.

### Best Practices

#### Messaging event are asynchronous

> Functions for messaging listens for messaging events asynchronously. As a consequence this can cause race conditions with other parts of the platform. Therefore, it is considered best practice to use bots instead of Functions for implementing routing logic. Functions is a good option to sync data with third-party systems like CRMs or to save data in the [Conversation Context Service](maven-context-warehouse-overview.html) in order to use it within Conversation Orchestrator. Routing via Functions makes sense whenever a conversation is in a stagnant state (i.e. not in process of being routed), e.g. a conversation is idle or a message line has been sent in off-hours. Otherwise, the asynchronous nature of its events might interfere with the proper flow of a "dynamic" conversation.

Reviewer: ```suggestion
Functions for messaging listens for messaging events asynchronously. As a consequence, this can cause race conditions with other parts of the conversation cloud. Therefore, it is considered best practice to use bots instead of Functions for implementing routing logic. Functions are a good option to enrich data with third-party systems like CRMs or to save data in the [Conversation Context Service](maven-context-warehouse-overview.html) in order to make it usable by another system like the Conversation Orchestrator. Routing via Functions makes sense whenever a conversation is in a stagnant state (i.e. not in process of being routed), e.g. a conversation is idle or a message line has been sent in off-hours. Otherwise, the asynchronous nature of its events might interfere with the proper flow of a "dynamic" conversation.
```

> <div class="important">if the authenticated consumer's previous conversation was auto-closed, and the new one opened within 48 hours of that, the new conversation event won't be triggered.</div>

Reviewer: notice

> #### Controller Bot will Retry on failed invocation

Reviewer: ```suggestion
#### Automatic Messages Error Handling
```

> If your function invocation takes too long or you have coding errors, the Controller Bot will retry up to 3 times, this will increase the number of invocations when your function does not perform well and could lead to more overload.

Reviewer: Never use controller bot...

> <div class="important">Controller Bot will retry up to 3 times when an invocation fails.</div>

Reviewer: Avoid div, just make it **big**

#### Multiple Messaging events are mapped to the same Functions event

> Since multiple conversational events are mapped to the same Functions event, you can differentiate the conversational event by the `cbotEventType` property in the invocation payload and adapt your for the interested events.

Reviewer: This is something you mention in the chapter where you speak about the mapping

Author: removed

> #### Any automatic message is disabled

Reviewer: This should not be a subtitel. It is something you can fit into the main text. Cutting down on the relevant information, that functions will be invoked even when auto message(s) are disabled. In order to completely avoid firing of functions you can either a or b

Author: I added it at the top in the warning when I mention this the first time,
